File: TheBookLore/app/src/main/java/com/penguin/thebooklore/ui/adapter/DashboardImagesRecyclerViewAdapter.java
package com.penguin.thebooklore.ui.adapter;

import android.content.Context;
import android.text.Layout;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.library.baseAdapters.BR;
import androidx.recyclerview.widget.RecyclerView;

import com.penguin.thebooklore.R;
import com.penguin.thebooklore.databinding.RecyclerItemDashboardImageBinding;
import com.penguin.thebooklore.model.WebImage;

import java.util.List;

public class DashboardImagesRecyclerViewAdapter extends RecyclerView.Adapter<DashboardImagesRecyclerViewAdapter.DashboardImageViewHolder> {

    private List<String> listImages;
    private Context context;

    public DashboardImagesRecyclerViewAdapter(Context context, List<String> listImages) {
        this.listImages = listImages;
        this.context = context;
    }

    @NonNull
    @Override
    public DashboardImagesRecyclerViewAdapter.DashboardImageViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        RecyclerItemDashboardImageBinding binding = DataBindingUtil.inflate(LayoutInflater.from(context), R.layout.recycler_item_dashboard_image, parent, false);

        return new DashboardImageViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull DashboardImagesRecyclerViewAdapter.DashboardImageViewHolder holder, int position) {
        String imageUrl = listImages.get(position);
        holder.bind(imageUrl);
    }

    @Override
    public int getItemCount() {
        if (listImages != null) {
            return listImages.size();
        } else {
            return 0;
        }
    }


    public class DashboardImageViewHolder extends RecyclerView.ViewHolder {

        public RecyclerItemDashboardImageBinding binding;

        public DashboardImageViewHolder(RecyclerItemDashboardImageBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(Object imageUrl) {
            binding.setVariable(BR.imageUrl, imageUrl);
            binding.executePendingBindings();
        }

    }
}

File: TheBookLore/app/src/main/java/com/penguin/thebooklore/ui/adapter/DashboardImagesRecyclerViewAdapter.kt
package com.penguin.thebooklore.ui.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.databinding.library.baseAdapters.BR
import androidx.recyclerview.widget.RecyclerView

import com.penguin.thebooklore.R
import com.penguin.thebooklore.databinding.RecyclerItemDashboardImageBinding
import com.penguin.thebooklore.model.Artwork

class DashboardImagesRecyclerViewAdapter(private val context: Context, private val listArtwork: List<Artwork>, private val clickListenerOpenArtDetail: (Artwork) -> Unit):
        RecyclerView.Adapter<DashboardImagesRecyclerViewAdapter.DashboardImageViewHolder>() {

    lateinit var binding: RecyclerItemDashboardImageBinding

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DashboardImageViewHolder {
        binding = DataBindingUtil.inflate(LayoutInflater.from(context),
                R.layout.recycler_item_dashboard_image,
                parent,
                false)

        return DashboardImageViewHolder(binding)
    }

    override fun onBindViewHolder(holder: DashboardImageViewHolder, position: Int) {
        holder.bind(listArtwork[position], clickListenerOpenArtDetail)
    }

    override fun getItemCount(): Int {
        return listArtwork.size
    }


    inner class DashboardImageViewHolder(var binding: RecyclerItemDashboardImageBinding) : RecyclerView.ViewHolder(binding.root) {

        fun bind(artwork: Artwork, clickListener: (Artwork) -> Unit) {
            binding.setVariable(BR.artwork, artwork)
            binding.setOnClickListenerOpenDetailArtwork { clickListener(artwork) }
            binding.executePendingBindings()
        }

    }
}
